File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {
        Arrays.sort(rocks);

        int left = 1;
        int right = distance;
        int answer = 0;

        while (left <= right) {
            int mid = (left + right) / 2;

            int removed = 0;
            int prev = 0; // 시작점

            for (int rock : rocks) {
                if (rock - prev < mid) {
                    removed++; // 제거
                } else {
                    prev = rock; // 유지
                }
            }

            // 마지막 구간 (도착지점)
            if (distance - prev < mid) {
                removed++;
            }

            if (removed > n) {
                // 너무 많이 제거 → 거리 너무 큼
                right = mid - 1;
            } else {
                // 가능 → 더 크게 시도
                answer = mid;
                left = mid + 1;
            }
        }

        return answer;
    }
}

File: 프로그래머스/4/43236. 징검다리/징검다리.js
function solution(distance, rocks, n) {
    // 바위 위치 오름차순 정렬
    rocks.sort((a, b) => a - b);

    let left = 1;
    let right = distance;
    let answer = 0;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        let removed = 0;
        let prev = 0; // 시작점

        for (const rock of rocks) {
            if (rock - prev < mid) {
                // 현재 바위 제거
                removed++;
            } else {
                // 현재 바위 유지
                prev = rock;
            }
        }

        // 마지막 구간(마지막 바위 ~ 도착지)
        if (distance - prev < mid) {
            removed++;
        }

        if (removed > n) {
            // 제거해야 하는 바위가 너무 많음 → 거리 줄이기
            right = mid - 1;
        } else {
            // 가능 → 최소 거리를 더 크게 시도
            answer = mid;
            left = mid + 1;
        }
    }

    return answer;
}